Keen/Dataset/Datasets.cs

```diff
@@ -0,0 +1,301 @@
+                .GetAsync(this.GetDatasetUrl(datasetName), _masterKey)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinition>(responseString, this.GetSerializerSettings());
+        }
+
+        public async Task<DatasetDefinitionCollection> ListDefinitions(int limit = 10, string afterName = null)
+        {
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var datasetResultsUrl = $"{_cachedDatasetRelativeUrl}?limit={limit}";
+
+            if (!string.IsNullOrWhiteSpace(afterName))
+            {
+                datasetResultsUrl += $"&after_name={afterName}";
+            }
+
+            var responseMessage = await _keenHttpClient
+                .GetAsync(datasetResultsUrl, _masterKey)
+                .ConfigureAwait( false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(false);
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinitionCollection>(responseString, GetSerializerSettings());
+        }
+
+        public async Task<IEnumerable<DatasetDefinition>> ListAllDefinitions()
+        {
+            var allDefinitions = new List<DatasetDefinition>();
+
+            var firstSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT);
+
+            if (firstSet?.Datasets == null)
+            {
+                throw new KeenException("Failed to fetch definition list");
+            }
+
+            if (!firstSet.Datasets.Any())
+            {
+                return allDefinitions;
+            }
+                
+            if (firstSet.Count <= firstSet.Datasets.Count())
+            {
+                return firstSet.Datasets;
+            }
+            
+            allDefinitions.AddRange(firstSet.Datasets);
+
+            do
+            {
+                var nextSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT, allDefinitions.Last().DatasetName);
```

Comment:
This doesn't seem right. It looks like in the best case scenario, if there are <100 datasets cached, we'll at least dispatch a GET to retrieve no datasets for no reason, and in worst case, we'll throw in the next few lines because we'll get back nothing. I may be missing something, but it seems this can only work if there are >= 101 dataset definitions to retrieve. We should instead be careful to only request another page if it's actually necessary.

Reply:
However, the `ListAllDefinitions_Success()` test uses `ListDatasetDefinitions.json` which doesn't have 100 datasets, so I'll have to take a closer look.

Reply:
I see, I misread the `if (firstSet.Count <= ...)` check, which should handle this situation.